File: Questions & Solutions/Balanced Binary Tree.cpp
/*
I used a bottom-up approach to check if the
tree is balanced. I computed the height of
each subtree and returned -1 if unbalanced,
allowing early termination when an imbalance
is detected.
O(n) time O(h) space
*/

class Solution {
    int height(TreeNode* node) {
        if (!node) return 0;
        int lh = height(node->left);
        if (lh == -1) return -1;     
        int rh = height(node->right);
        if (rh == -1) return -1;        
        if (abs(lh - rh) > 1) return -1;   
        return 1 + max(lh, rh);            
    }
public:
    bool isBalanced(TreeNode* root) {
        return height(root) != -1;
    }
};

File: Questions & Solutions/Balanced Binary Tree.py
"""
  Use DFS to traverse the tree and check balance at each node.
  Each node returns a pair: [is_balanced, height].
  A tree is balanced if: left subtree is balanced AND right subtree is balanced AND height difference <= 1.
  Return the height as 1 + max(left height, right height) for parent nodes to use.
  Base case: empty node returns [True, 0].
  O(n) time, O(h) space where h is tree height for recursion stack.
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(root):
            if not root:
                return [True, 0]

            left, right = dfs(root.left), dfs(root.right)
            balanced = left[0] and right[0] and abs(left[1] - right[1]) <= 1
            return [balanced, 1 + max(left[1], right[1])]

        return dfs(root)[0]